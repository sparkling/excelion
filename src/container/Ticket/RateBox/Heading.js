import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {Type} from './index';

const StyledHeading = styled.div `
  display: flex;
  flex-direction: ${props => props.buySell === Type.BUY ? "row-reverse" : "row"};
  padding: 0.5em 0.75em;
  & > *:${props => props.buySell === Type.BUY ? "last" : "first"}-child {
    flex: 1;
  }
`

const MostSignificant = styled.div `
  font-size: 0.8em;
`
MostSignificant.displayName = 'MostSignificant';

const Action = styled.div `
  font-size: 0.8em
`
Action.displayName = 'Action';

const Heading = ({ rateMostSignificantDigits, buySell, currencySymbol }) =>
  (<StyledHeading buySell={buySell} >
    <Action>{buySell} {currencySymbol}</Action>
    <MostSignificant>{rateMostSignificantDigits}</MostSignificant>
  </StyledHeading>);

Heading.propTypes = {
  rateMostSignificantDigits: PropTypes.string.isRequired,
  buySell: PropTypes.string.isRequired,
  currencySymbol: PropTypes.string.isRequired,
}

export default Heading;
