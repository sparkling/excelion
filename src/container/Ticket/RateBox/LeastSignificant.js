import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const StyledLeastSignificant = styled.div `
  text-align: center;
`

const Digit = styled.span `
  font-size: 5em;
  &:last-child {
    font-size: 2em;
  }
`
Digit.displayName='Digit';

const LeastSignificant = ({ rateLeastSignificantDigits }) => {
  const digits = rateLeastSignificantDigits.split("").map((digit, idx) => {
      return <Digit key={idx}>{digit}</Digit>;
    });

  return (<StyledLeastSignificant>{digits}</StyledLeastSignificant>);
}

LeastSignificant.propTypes = {
  rateLeastSignificantDigits: PropTypes.string.isRequired
}

export default LeastSignificant;
