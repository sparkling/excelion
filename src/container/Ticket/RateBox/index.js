import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Heading from './Heading';
import LeastSignificant from './LeastSignificant';

export const Type = {
  "BUY": "BUY",
  "SELL": "SELL"
}

const Box = styled.div `
  font-family: arial;
  flex: 1;
  color: white;
  background-color: ${props => props.bgColor || "grey"};
`
Box.displayName = 'Box';

class RateBox extends React.PureComponent {

  constructor(props) {
    super(props);
    this.state = {
      bgColor: "grey",
    };
  }

  componentDidUpdate(prevProps) {
    const prevValue = parseInt(prevProps.rateLeastSignificantDigits, 10);
    const nextValue = parseInt(this.props.rateLeastSignificantDigits, 10);

    if (prevValue > nextValue) {
      this.setDecrease();
    }
    else if (prevValue < nextValue) {
      this.setIncrease();
    }
  }

  setIncrease() {
    this.setState(() => {
      return { bgColor: "blue"}
    });
  }

  setDecrease() {
    this.setState(() => {
      return { bgColor: "red"}
    });
  }

  render () {
    return (
      <Box bgColor={this.state.bgColor} onClick={this.props.onClick}>
        <Heading
          buySell={this.props.buySell}
          currencySymbol={this.props.currencySymbol}
          rateMostSignificantDigits={this.props.rateMostSignificantDigits}
        />
        <LeastSignificant
          rateLeastSignificantDigits={this.props.rateLeastSignificantDigits}
        />
      </Box>
    );
  }
}

RateBox.propTypes = {
  onClick: PropTypes.func.isRequired,
  rateMostSignificantDigits: PropTypes.string.isRequired,
  rateLeastSignificantDigits: PropTypes.string.isRequired,
  currencySymbol: PropTypes.string.isRequired,
  buySell: PropTypes.string.isRequired,
}

export default RateBox;
