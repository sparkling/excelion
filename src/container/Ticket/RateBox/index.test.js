import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { mount } from 'enzyme';
import RateBox, {Type} from './index.js';

const defaultRateLeastSignificantDigits = "841"
const defaultRateMostSignificantDigits = "1.40"
const defaultCurrencySymol = "EUR";
const defaultBuySell = Type.BUY;
const defaultMockCallBack = jest.fn();

const defaultMount = mount(
  <RateBox
    rateMostSignificantDigits={defaultRateMostSignificantDigits}
    rateLeastSignificantDigits={defaultRateLeastSignificantDigits}
    currencySymbol={defaultCurrencySymol}
    buySell={defaultBuySell}
    onClick={defaultMockCallBack}
   />);

describe('<RateBox />', () => {

  it('should render the same BUY snapshot', () => {
    const tree = renderer.create(<RateBox
      rateMostSignificantDigits="1.40"
      rateLeastSignificantDigits="872"
      currencySymbol="EUR"
      buySell={Type.BUY}
      onClick={()=>{}}
     />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render the same SELL snapshot', () => {
    const tree = renderer.create(<RateBox
      rateMostSignificantDigits="1.40"
      rateLeastSignificantDigits="872"
      currencySymbol="EUR"
      buySell={Type.SELL}
      onClick={()=>{}}
     />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render an action description', () => {
    expect(defaultMount.find('Action').text()).toBe(`${defaultBuySell} ${defaultCurrencySymol}`);
  });

  it('should render most significant digits', () => {
    expect(defaultMount.find('MostSignificant').text()).toBe(defaultRateMostSignificantDigits);
  });

  it('should render least significant digits', () => {
    expect(defaultMount.find('LeastSignificant').render().text()).toBe(defaultRateLeastSignificantDigits);
  });

  it('should capture a click', () => {
    defaultMount.simulate('click');
    expect(defaultMockCallBack.mock.calls.length).toEqual(1);
  });

  it('should change background to red when rate is lowered', () => {
    const value = parseInt(defaultRateLeastSignificantDigits, 10);
    defaultMount.setProps({ rateLeastSignificantDigits: (value - 1).toString() });
    defaultMount.update();
    expect(defaultMount.find('Box').prop('bgColor')).toEqual('red');
  });

  it('should change background to blue when rate is increased', () => {
    const value = parseInt(defaultRateLeastSignificantDigits, 10);
    defaultMount.setProps({ rateLeastSignificantDigits: (value + 1).toString() });
    defaultMount.update();
    expect(defaultMount.find('Box').prop('bgColor')).toEqual('blue');
  });

});
