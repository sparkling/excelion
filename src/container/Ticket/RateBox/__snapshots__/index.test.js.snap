// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RateBox /> should render the same BUY snapshot 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
  padding: 0.5em 0.75em;
}

.c1 > *:last-child {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 {
  font-size: 0.8em;
}

.c2 {
  font-size: 0.8em;
}

.c4 {
  text-align: center;
}

.c5 {
  font-size: 5em;
}

.c5:last-child {
  font-size: 2em;
}

.c0 {
  font-family: arial;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  color: white;
  background-color: grey;
}

<div
  className="c0"
  onClick={[Function]}
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      BUY
       
      EUR
    </div>
    <div
      className="c3"
    >
      1.40
    </div>
  </div>
  <div
    className="c4"
  >
    <span
      className="c5"
    >
      8
    </span>
    <span
      className="c5"
    >
      7
    </span>
    <span
      className="c5"
    >
      2
    </span>
  </div>
</div>
`;

exports[`<RateBox /> should render the same SELL snapshot 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding: 0.5em 0.75em;
}

.c1 > *:first-child {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 {
  font-size: 0.8em;
}

.c2 {
  font-size: 0.8em;
}

.c4 {
  text-align: center;
}

.c5 {
  font-size: 5em;
}

.c5:last-child {
  font-size: 2em;
}

.c0 {
  font-family: arial;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  color: white;
  background-color: grey;
}

<div
  className="c0"
  onClick={[Function]}
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      SELL
       
      EUR
    </div>
    <div
      className="c3"
    >
      1.40
    </div>
  </div>
  <div
    className="c4"
  >
    <span
      className="c5"
    >
      8
    </span>
    <span
      className="c5"
    >
      7
    </span>
    <span
      className="c5"
    >
      2
    </span>
  </div>
</div>
`;
